pkg/controller.v1/mpi/mpijob_controller.go

```diff
@@ -0,0 +1,1402 @@
+		UpdateFunc: util.OnDependentUpdateFunc(&r.JobController),
+		DeleteFunc: util.OnDependentDeleteFunc(r.Expectations),
+	}); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+//mpijob not need delete services
+func (r *MPIJobReconciler) DeletePodsAndServices(runPolicy *commonv1.RunPolicy, job interface{}, pods []*corev1.Pod) error {
+	if len(pods) == 0 {
+		return nil
+	}
+
+	// Delete nothing when the cleanPodPolicy is None.
+	if *runPolicy.CleanPodPolicy == commonv1.CleanPodPolicyNone {
+		return nil
+	}
+
+	for _, pod := range pods {
+		// Note that pending pod will turn into running once schedulable,
+		// not cleaning it may leave orphan running pod in the future,
+		// we should treat it equivalent to running phase here.
+		if *runPolicy.CleanPodPolicy == commonv1.CleanPodPolicyRunning && pod.Status.Phase != corev1.PodRunning && pod.Status.Phase != corev1.PodPending {
+			continue
+		}
+		if err := r.PodControl.DeletePod(pod.Namespace, pod.Name, job.(runtime.Object)); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// reconcileServices checks and updates services for each given ReplicaSpec.
+// It will requeue the job in case of an error while creating/deleting services.
+// mpijob not need services
+func (jc *MPIJobReconciler) ReconcileServices(
+	job metav1.Object,
+	services []*corev1.Service,
+	rtype commonv1.ReplicaType,
+	spec *commonv1.ReplicaSpec) error {
+	return nil
+}
+
+func (r *MPIJobReconciler) ControllerName() string {
+	return controllerName
+}
+
+func (r *MPIJobReconciler) GetAPIGroupVersionKind() schema.GroupVersionKind {
+	return mpiv1.GroupVersion.WithKind(mpiv1.Kind)
+}
+
+func (r *MPIJobReconciler) GetAPIGroupVersion() schema.GroupVersion {
+	return mpiv1.GroupVersion
+}
+
+func (r *MPIJobReconciler) GetGroupNameLabelValue() string {
+	return mpiv1.GroupVersion.Group
+}
+
+// Same as Func (tc *TFController) SetClusterSpec(...) in pod.go
+func (r *MPIJobReconciler) SetClusterSpec(job interface{}, podTemplate *corev1.PodTemplateSpec, rtype, index string) error {
+	return nil
+}
+
+func (r *MPIJobReconciler) GetDefaultContainerName() string {
+	return mpiv1.DefaultContainerName
+}
+
+func (r *MPIJobReconciler) GetDefaultContainerPortName() string {
+	return mpiv1.DefaultPortName
+}
+
+func (r *MPIJobReconciler) IsMasterRole(replicas map[commonv1.ReplicaType]*commonv1.ReplicaSpec,
+	rtype commonv1.ReplicaType, index int) bool {
+	return false
+}
+
+func (r *MPIJobReconciler) GetJobFromInformerCache(namespace, name string) (metav1.Object, error) {
+	mpijob := &mpiv1.MPIJob{}
+	err := r.Get(context.Background(), types.NamespacedName{
+		Namespace: namespace, Name: name,
+	}, mpijob)
+	return mpijob, err
+}
+
+// onOwnerCreateFunc modify creation condition.
+func (r *MPIJobReconciler) onOwnerCreateFunc() func(event.CreateEvent) bool {
+	return func(e event.CreateEvent) bool {
+		mpiJob, ok := e.Object.(*mpiv1.MPIJob)
+		if !ok {
+			return true
+		}
+
+		r.Scheme.Default(mpiJob)
```

Comment:
I'm not convinced that this is safe. If a user changes an MPIJob before the controller had a chance to update the job, I think the defaults would be cleared. And maybe there could be a race condition in the client cache, but I'm not sure how the cache is implemented in controller-runtime.

cc @Jeffwan who did the same for tf-operator

Reply:
@alculquicondor 
I did same thing for tf.  Currently, it's still using a lightweight solution, the ideal way is to move the logics to webhooks.  If user changes the job, the event sequence for the same object is guaranteed but since they are different kind of events and handled by different methods, I would say there's chance to get into the race condition. 

T1: onOwnerCreateFunc is revoked and it tries to set default `RunPolicy`
T2: onUpdate() is revoked and it updates similar fields
T3. In the end, `onOwnerCreateFunc` set defaults.

Honestly, I have not played with this case. I agree even if we won't update that fast but the case is still there theoretically. Is there a better way to handle this without webhook?  Maybe we can move most of logics to `Defaulting` https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#defaulting